lang/java/reef-io/src/main/java/org/apache/reef/io/network/impl/NetworkConnectionServiceImpl.java

```diff
@@ -231,9 +285,20 @@ public Identifier getNetworkConnectionServiceId() {
 
   @Override
   public void close() throws Exception {
-    LOG.log(Level.FINE, "Shutting down");
-    this.nameServiceRegisteringStage.close();
-    this.nameServiceUnregisteringStage.close();
-    this.transport.close();
+    if (isClosed.compareAndSet(false , true)) {
+      LOG.log(Level.FINE, "Shutting down");
+      this.nameServiceRegisteringStage.close();
+      this.nameServiceUnregisteringStage.close();
```

Comment:
We are closing the stage here, then using it below in `unregisterAllConnectionFactories()`. The order should be reversed.